### DISCUSSION TITLE
FFmpeg Hardware Acceleration with vaapi

### DISCUSSION THREAD

**Discussion by Nycrera** [bytedeco/javacv#1867] [Q&A] [answered]
Hey, I was wondering how should I set the codec for a FFmpegFrameRecorder to get hardware accelerated encoding with va-api on Linux:
Here is my current setup:
```java
recorder = new FFmpegFrameRecorder("rtp://127.0.0.1:1234", videoGrabber.getImageWidth(),
				videoGrabber.getImageHeight(), audioGrabber.getAudioChannels());
		recorder.setFormat("rtp_mpegts");

		recorder.setOption("hwaccel", "vaapi");
		recorder.setOption("hwaccel_device", "/dev/dri/renderD128");
		recorder.setOption("hwaccel_output_format", "vaapi");
		recorder.setOption("vf", "scale_vaapi=format=nv12"); // ???
		
		                               //recorder.setVideoCodec(avcodec.AV_CODEC_ID_H265);
		recorder.setVideoCodecName("hevc_vaapi");
		recorder.setAudioCodec(avcodec.AV_CODEC_ID_AAC);
		recorder.setSampleRate(audioGrabber.getSampleRate());
		recorder.setVideoBitrate(1 * 1000 * 1000);
		recorder.setAudioBitrate(48 * 1000);
		recorder.setFrameRate(30);
```
Now, I couldn't find any codec id definitions in javacv with vaapi (Maybe I'm just missing it.) so I went with setting the codec name. However, I am not sure if this is the correct way to do this.

Anyhow, the error I am facing right now is:
`[hevc_vaapi @ 0x7fd710c24140] Specified pixel format yuv420p is invalid or not supported`

However despite trying multiple different pixel formats with `recorder.setPixelFormat()` I can't get any of them to work.
Also, this was never a problem when FFmpeg was used with the same encoder and pixel format.

For the record here is the same thing in cli working:
`ffmpeg -re -f v4l2 -framerate 30/1 -input_format mjpeg -video_size 1280x720 -hwaccel vaapi -hwaccel_device /dev/dri/renderD128 -hwaccel_output_format vaapi -i /dev/video0 -f alsa -i hw:0 -map 0:v -map 1:a -vf 'scale_vaapi=format=nv12' -c:a aac -b:a 48K -c:v hevc_vaapi -b:v 1M -f rtp_mpegts rtp://127.0.0.1:1234`

**Comment by saudet** [comment, marked as answer]
It looks like the only pixel format that hevc_vaapi supports is AV_PIX_FMT_VAAPI:
https://github.com/FFmpeg/FFmpeg/blob/master/libavcodec/vaapi_encode_h265.c#L1463
So try to call setPixelFormat(AV_PIX_FMT_VAAPI) and see what that gives.

**Reply by Nycrera** [reply]
Yes this seems to be the case!

**Reply by Nycrera** [reply]
Hey @saudet, sorry if bothering but, when I am encoding my video just like in the previous code. I am getting an error:
`[hevc_vaapi @ 0x7fdb08bdafc0] A hardware frames reference is required to associate the encoding device`

A quick google search revealed that when using ffmpeg with vaapi, [an environment variable called LIBVA_DRIVER_NAME=iHD, (in my case iHD) needs to define which driver to use.](https://forum.shotcut.org/t/hardware-encoding-linux-radeon-navi10-5600xt/22943) And this information is correct as I was setting this variable while I was encoding on the command line with ffmpeg (it was working).

Now, I am suspecting that javacv's integral ffmpeg functions does not see this environment variable that I have exported on my command line. What do you think? How should I have defined this environment variable?

Or is this totally unrelated?

**Reply by saudet** [reply]
It sounds like a variable that gets read by the driver, and if that's the case, it doesn't matter which program you're running. Something else is missing from the configuration.

**Reply by Nycrera** [reply]
I have found the error line in ffmpeg as:
[https://github.com/FFmpeg/FFmpeg/blob/11912f65ef5ad12db938e404eafc855af1bc3859/libavcodec/vaapi_encode.c#L2555](https://github.com/FFmpeg/FFmpeg/blob/11912f65ef5ad12db938e404eafc855af1bc3859/libavcodec/vaapi_encode.c#L2555)
so it looks like the vaapi can't initialize.

**Reply by saudet** [reply]
@b005t3r Are you having any problems with VA-API on Linux?

**Reply by b005t3r** [reply]
> @b005t3r Are you having any problems with VA-API on Linux?

I don't think I ever got to use it. I think it wasn't working at the moment when I tried it, but I see things hasn't improved since :)

**Reply by saudet** [reply]
Thanks to @silenium-dev, this has been fixed! Please give it a try with the snapshots: http://bytedeco.org/builds/
